### PR TITLE
feat: support unsubscribing by weakmap of handlers and wrapped handlers

### PR DESCRIPTION
**Pull request by deloreyj**

### CODE REVIEW

File: src/MagentoDataLayerBase.ts [hunk 1]
```diff
@@ -20,19 +20,30 @@ export abstract class MagentoDataLayerBase {
...
-    handler: (event: Record<string, unknown>) => void
+    handler: MagentoDataLayerEventHandler
   ): void {
+    if (!this.mdl._handlerMapper.get(handler)) {
+      if (
+        process.env.NODE_ENV === "development" ||
+        process.env.NODE_ENV === "test"
+      )
+        // eslint-disable-next-line no-console
+        console.log("handlers must be registered before they are removed");
```
Comment (rancollier) [inline src/MagentoDataLayerBase.ts:41]: what about console.error() or console.warn() to elevate the visual status to the user. Unless this is a standard.

File: src/MagentoDataLayerBase.ts [hunk 2]
```diff
@@ -20,19 +20,30 @@ export abstract class MagentoDataLayerBase {
     name: EventName,
     handler: MagentoDataLayerEventHandler
   ): void {
+    const wrappedHandler = (event: EventName) => handler(event, this.mdl);
+
+    this.mdl._handlerMapper.set(handler, wrappedHandler);
```
Comment (bradgarropy) [inline src/MagentoDataLayerBase.ts:25]: Shouldn't this map the `event` to the `wrappedHandler`? Or am I misunderstanding something?

```javascript
this.mdl._handlerMapper.set(event, wrappedHandler);
```

File: src/MagentoDataLayerBase.ts [hunk 3]
```diff
@@ -20,19 +20,30 @@ export abstract class MagentoDataLayerBase {
     name: EventName,
     handler: MagentoDataLayerEventHandler
   ): void {
+    const wrappedHandler = (event: EventName) => handler(event, this.mdl);
```
Comment (bradgarropy) [inline src/MagentoDataLayerBase.ts:23]: Yes, this is key right here!

File: src/MagentoDataLayerBase.ts (hunk 2 above)
Comment (deloreyj) [reply src/MagentoDataLayerBase.ts:25]: It should map handler to wrappedHandler. An event can have many subscribers. When a user wants to unregister a handler they have to pass in the handler function they originally passed in so the ACDL knows which handler to unregister. The problem was the handler they passed in to add the event listener isn't the same one I pass on to the ACDL. So I have to maintain a mapping between the handler they pass in and the wrappedHandler I pass along.

**Pull request merged**